### PR TITLE
Pass input tensor __dict__ along to placeholder nodes

### PR DESCRIPTION
**Pull request by voznesenskym**
```
import torch
import torch.nn as nn

import torch._dynamo.config
import torch._inductor.config

def pre_attention_state_ops(input, mems, state):
    lc_key = state[0]
    lc_val = state[1]
    bar = []
    for i in range(0, 4):
        bar2 = []
        for j in range(0, 3):
            bar2.append(
                lc_key + lc_val + torch.tensor([0.1, 0.25, 0.4, 0.5, 0.1])
            )
        bar.append(bar2)

    return bar

mems = torch.tensor([[[1.8364, 0.2724, -1.4917, -0.4367, 0.8640]]])
state = [
    torch.tensor([[[1.0517, 0.3848, -0.6472, 0.0823, 0.9116]]]),
    torch.tensor([[[1.0517, 0.3848, -0.6472, 0.0823, 0.9116]]]),
]
i = torch.tensor(
    [
        [0.0313, -0.1487, -0.3846, -0.5321],
        [-1.7073, 1.3331, -0.0890, -1.4935],
        [-0.8314, -0.1862, -0.5935, 1.5232],
    ]
)

torch._dynamo.tag(mems, "MEMS")
torch._dynamo.tag(i, "FOO")
torch._dynamo.tag(state[0], "STATE_0")
torch._dynamo.tag(state[1], "HMMM")

exported = torch._dynamo.export(pre_attention_state_ops, i, mems, state)
out_graph = exported[0]

dynamo_result = out_graph(i, mems, state)
nodes = list(out_graph.graph.nodes)
placeholders = [node for node in nodes if node.op == "placeholder"]
for placeholder in placeholders:
    if "tags" in placeholder.meta:
        print("PLACEHOLDER TAGS?", placeholder.meta["tags"])
        
```

prints 

PLACEHOLDER TAGS? ['STATE_0']
PLACEHOLDER TAGS? ['HMMM']

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by jansel** [comment]
Debugging tools seem great, but I don't like the UX of `torch._dynamo.tag()`.
1) For this specific example (where the tags are just variable names) perhaps it would be better for dynamo to just capture variable names in its debugging metadata.
2) For other uses of tagging, could we design something that worked in eager mode too?  I don't see why tensor labels/tags need to be dynamo specific.

**Comment by voznesenskym** [comment]
> Debugging tools seem great, but I don't like the UX of `torch._dynamo.tag()`.
> 
> 1. For this specific example (where the tags are just variable names) perhaps it would be better for dynamo to just capture variable names in its debugging metadata.

@jansel that's just an example :P I think @suo wants to use it for tracking arbitrary metadata on tensors to provide hints for backends. Specifically, deferred initialization is one he needs for a current project. 

> 2. For other uses of tagging, could we design something that worked in eager mode too?  I don't see why tensor labels/tags need to be dynamo specific.

Sure! I wanted to contain this to just the current request export, but would be open to other ideas.

That being said, tensor tagging could be generally useful, but would need more design. One reason I shied away from it is specifically because reasoning about the propagation of a tag can become tricky around copies, aliases, etc. Frankly, I don't think I have enough experience here to make good calls.

Ex: One could argue that a copy should copy the tag, one could argue that identity of the new tensor differs, and so should not be tagged (it's not the exact same tensor the user tagged). Does an intermediate tensor's tag show its lineage? in `a = b + c`, is `a` tagless, or does it get the tags of `b` and `c`? 

I also worry about tagging being extended into systems in general, in my experience outside of pytorch, tagging systems that more than just a shallow one (like in this PR) end up going from debugging, to analytics/counting, to eventually acting as critical input for some control flow. It feels a bit like a pandora's box, where someone will eventually end up building something that checks `if tensor.tag == foo`, and thats not really a great API surface imo. Putting it on dynamo feels like it lets us work around this whole set of concerns but purposely letting it start off as a rather anemic, simple, thing.

cc: @gchanan @ezyang as we touched this briefly

**Comment by ezyang** [comment]
An "eager agnostic" choice would be to save exactly what is on the `__dict__` of input tensors, as specified by what we save when those tensors get pickled. Not really sure how well this would work in practice.

**Comment by voznesenskym** [comment]
> An "eager agnostic" choice would be to save exactly what is on the `__dict__` of input tensors, as specified by what we save when those tensors get pickled. Not really sure how well this would work in practice.

Would this save the tensor values?

**Comment by ezyang** [comment]
No.

**Comment by voznesenskym** [comment]
> No.

Hmm, you still need the placeholder association, but just using the entire dict seems reasonable.

**Comment by ezyang** [comment]
> For other uses of tagging, could we design something that worked in eager mode too? I don't see why tensor labels/tags need to be dynamo specific.

But @jansel, do you really want to export tags on intermediate nodes of the graph? This seems difficult to do in general because FX nodes are not necessarily in one-to-one correspondence with Python tensors. And if you don't care about propagating tags, then I don't really see what "working eager mode" means; you can just slap stuff in `__dict__` and call it a day.

**Comment by jansel** [comment]
I'd be fine with just copying stuff from `tensor.__dict__` into `plaeholder_node.meta["tensor_dict"]`.

**Comment by gchanan** [comment]
the dict solution sounds good to me!

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -878,6 +878,8 @@ def _clone_input(value):
         specialized_props["specialized_value"] = specialized_value
 
         options.update(specialized_props)
+        assert "tensor_dict" not in proxy.node.meta
+        proxy.node.meta["tensor_dict"] = example_value.__dict__
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:882]: You probably want to do a shallow copy

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -1002,6 +1004,9 @@ def wrap_to_fake_tensor_and_record(
                 source=source,
             )
         )
+        for k, v in e.__dict__.items():
+            if k not in fake_e.__dict__:
+                fake_e.__dict__[k] = v
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:1009]: This is a weird thing to do and it has the potential to corrupt fake tensor state if the user tensor redefines an attribute fake tensor uses internally.

**Review by ezyang — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/builder.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:1009]: hmm, I could make a weak mapping between real<>fake, I suppose, we just need a way to launder it through

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py]: I wasn't sure, I was actually thinking maybe a deep copy, in case this graph outlives the tensors, but I wasn't sure...

Comment (ezyang) [reply torch/_dynamo/variables/builder.py]: The main reason to shallow copy is so that modifying the dict doesnt spooky action at a distance between meta and the og tensor

**Comment by suo** [comment]
Chiming in to say: from the functionality perspective, this is precisely what I want out of the tensor tagging stuff. I vaguely would prefer only propagating a specific field (e.g. `tensor.tags = {"all": "the stuff"}`) instead of the `__dict__`, on the principle that it narrows the chance of unintentional propagation, but that is a minor/non-blocking concern.

**Comment by voznesenskym** [comment]
> I think it's fine to dump everything in as long as you're staying in the Python process. At the point you are actually dumping to a serialized format, you should be more deliberate about what you put in there, but if we're chucking stuff in FX node meta there isn't any expectation this all gets automatically serialized anyway, I think.

As long as there is not a pre-existing serialization expectation, or back compat contract. If there is, I would be willing to scope this back down to an explicit field, "tags" or something, or to go change serialization logic to be more defensive about the __dict__ contents.

Another point - __dict__ keeps this free and open to any value types, worth discussing if we want that.

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py]: sg - let's decide if we are keeping dict below, and if we are, I can rev this to add a copy

**Comment by ezyang** [comment]
I mean, so, does anyone here know what the GraphModule -> byte format pipeline looks like? That's the operant question here.

**Comment by suo** [comment]
An experimental version serializing to thrift is implemented (internally) [here](https://fburl.com/code/go7cse4p).

Thinking about it more, I am now more strongly in favor of a special `tags` field and restrictions on what can be put there (only strings). If we are going to be providing the serialization system—which we will be—that's the only way we can offer guarantees that tags will be properly serialized.

**Comment by voznesenskym** [comment]
> An experimental version serializing to thrift is implemented (internally) [here](https://fburl.com/code/go7cse4p).
> 
> Thinking about it more, I am now more strongly in favor of a special `tags` field and restrictions on what can be put there (only strings, or more precisely things that implement `__str__`). If we are going to be providing the serialization system—which we will be—that's the only way we can offer guarantees that tags will be properly serialized.

@jansel and @ezyang do you care if we go back to this from __dict__?

**Comment by ezyang** [comment]
The Thrift serializer hard codes to only serialize val from meta. It doesn't matter if we put more junk or not in meta, it won't get serialized; in particular, chucking `tensor_dict` into meta won't do anything.

OK, so how would we adjust the Thrift serializer? You have two choices:

1. Explicit configuration. You tell Thrift "I want to serialize the attributes tags on all placeholder tensors." This clearly must be a Thrift side configuration. It doesn't matter if you chuck `tensor_dict` into the meta; in the E2E, you'll just look for `meta['tensor_dict']['tags']` to actually serialize
2. Configuration by convention. You tell Thrift "I want to serialize all attributes on placeholder tensors that have property X (e.g., they are prefixed with `thrift_` or their type is string or their type is ExportMeta)". It still doesn't matter if you chuck `tensor_dict` into the meta; in the E2E Thrift will grovel through it to find things that match the convention.

Does this make sense?

**Comment by voznesenskym** [comment]
> The Thrift serializer hard codes to only serialize val from meta. It doesn't matter if we put more junk or not in meta, it won't get serialized; in particular, chucking `tensor_dict` into meta won't do anything.
> 
> 
> 
> OK, so how would we adjust the Thrift serializer? You have two choices:
> 
> 
> 
> 1. Explicit configuration. You tell Thrift "I want to serialize the attributes tags on all placeholder tensors." This clearly must be a Thrift side configuration. It doesn't matter if you chuck `tensor_dict` into the meta; in the E2E, you'll just look for `meta['tensor_dict']['tags']` to actually serialize
> 
> 2. Configuration by convention. You tell Thrift "I want to serialize all attributes on placeholder tensors that have property X (e.g., they are prefixed with `thrift_` or their type is string or their type is ExportMeta)". It still doesn't matter if you chuck `tensor_dict` into the meta; in the E2E Thrift will grovel through it to find things that match the convention.
> 
> 
> 
> Does this make sense?

Option 1 is far more thrift idiomatic. It also shouldn't change this PR, now that I think about it. 

There is debugging value in having the dict ride through to appear on placeholders, and there is also value is nominating a field as special (tags) to then teach serialization about. 

That can be done in a stacked PR on top of this one.

**Comment by suo** [comment]
I am fine with option1, which I interpret as morally the same (from the perspective of export) as only preserving `tensor.special_tag_attr` but threading the needle to have something nice for eager debugging purposes as well.

**Review by ezyang — APPROVED** [review]

**Review by jansel — APPROVED** [review]
